File: scripts/crypto-interop/verify_vector.py
"""
Interop test: JS -> Python.
Decrypts scripts/crypto-interop/vector-js.json (produced by
scripts/crypto-roundtrip.mjs) with the worker-side crypto module, then
produces vector-py.json for the Node side to decrypt.

Run: python scripts/crypto-interop/verify_vector.py
"""
import base64
import hashlib
import json
import os
import sys

sys.path.insert(0, os.path.join(os.path.dirname(__file__), "..", "..", "workers", "common"))

from worker_common.crypto import (  # noqa: E402
    PLAINTEXT_BLOCK,
    decrypt_bytes,
    encrypt_bytes,
    fingerprint_key,
)

HERE = os.path.dirname(__file__)


def main() -> None:
    with open(os.path.join(HERE, "vector-js.json"), encoding="utf-8") as fh:
        vector = json.load(fh)

    dek = base64.b64decode(vector["dek"])
    ciphertext = base64.b64decode(vector["ciphertext"])

    plain = decrypt_bytes(dek, vector["fileId"], vector["ownerId"], ciphertext)
    digest = hashlib.sha256(plain).hexdigest()
    assert digest == vector["plaintext_sha256"], (
        f"JS vector decryption mismatch: {digest} != {vector['plaintext_sha256']}"
    )
    assert fingerprint_key(dek) == vector["dek_fingerprint"], "fingerprint mismatch"
    print("✓ Python decrypts the JS vector (multi-frame) and fingerprints match")

    # Produce the Python -> JS vector.
    plain_py = (
        b"Nest2D interop vector, Python side.\n"
        + os.urandom(PLAINTEXT_BLOCK + 42)
    )
    encrypted_py = encrypt_bytes(dek, vector["fileId"], vector["ownerId"], plain_py)
    with open(os.path.join(HERE, "vector-py.json"), "w", encoding="utf-8") as fh:
        json.dump(
            {
                "dek": vector["dek"],
                "fileId": vector["fileId"],
                "ownerId": vector["ownerId"],
                "plaintext_sha256": hashlib.sha256(plain_py).hexdigest(),
                "ciphertext": base64.b64encode(encrypted_py).decode("ascii"),
                "dek_fingerprint": fingerprint_key(dek),
            },
            fh,
            indent=2,
        )
    print("✓ vector-py.json written")


if __name__ == "__main__":
    main()

File: scripts/crypto-interop/verify_vector.mjs
/**
 * Interop test: Python -> JS. Decrypts vector-py.json with the server-side
 * crypto module. Run after scripts/crypto-interop/verify_vector.py.
 *
 * Run: node scripts/crypto-interop/verify_vector.mjs
 */
import crypto from 'node:crypto'
import { readFileSync } from 'node:fs'

globalThis.useRuntimeConfig = () => ({ encryptionMasterKey: '' })
globalThis.createError = (opts) => Object.assign(new Error(opts.statusMessage), opts)

const { decryptBuffer, fingerprintKey } = await import('../../server/utils/crypto.js')

const vector = JSON.parse(readFileSync(new URL('./vector-py.json', import.meta.url), 'utf8'))
const dek = Buffer.from(vector.dek, 'base64')
const plain = decryptBuffer(dek, vector.fileId, vector.ownerId, Buffer.from(vector.ciphertext, 'base64'))
const digest = crypto.createHash('sha256').update(plain).digest('hex')

if (digest !== vector.plaintext_sha256) {
    throw new Error(`Python vector decryption mismatch: ${digest} != ${vector.plaintext_sha256}`)
}
if (fingerprintKey(dek) !== vector.dek_fingerprint) {
    throw new Error('fingerprint mismatch')
}
console.log('✓ Node decrypts the Python vector (multi-frame) and fingerprints match')
